Add what is right about modules (#5)
ctsit/redcap_external_module_development_guide: docs/guide_for_admins_and_devs.md

This embedded configuration made code sharing harder. If you wanted the new version of the hook code you had to apply your configuration code to the new code then deploy that. Even within your own systems, if your staging and production systems needed different configuration details you couldn't use the same code on both systems.

Another configuration method was to store the configuration in a record in the redcap_log_event table. It's a little kludgy, but it worked. Yet when an extension that used log-event table method of configuration management failed, it was hard to debug and even harder to fix.

Code that was only supposed to be applied to a few projects presented further challenges. The methods for managing a per-project deployment were in the hands of the developer or system admin instead of the REDCap Admin.

> When hooks worked right, they work great, but a PHP error in a hook or plugin can be unforgiving. With _PHP errors_ configuration option  turned off, a production system displays nothing but an empty white screen if the code contains a PHP syntax error. A buggy plugin could be toxic to a REDCap system. Reverting the code of a buggy plugin or hook was as tedious as the deployment.

Reviewer: ```suggestion
When hooks worked right, they worked great, but a PHP error in a hook or plugin could be unforgiving. With _PHP errors_ configuration option  turned off, a production system displayed nothing but an empty white screen if the code contains a PHP syntax error. A buggy plugin could be toxic to a REDCap system. Reverting the code of a buggy plugin or hook was as tedious as the deployment.
```
More consistent application of past tense

Author: I agree with the first fix in the line, but not the second.  "...a production system displays nothing but..." is still accurate. Could you revert that bit?

Reviewer: I updated it in place

## REDCap Repo

Yet the EM Framework does not stand alone. Many of the features it provides are backed by the _REDCap Repo_. The REDCap Repo is a public collection of vetted REDCap modules maintained by the Vanderbilt UMC REDCap team. These modules have been submitted by the members of the REDCap community to be sharing within the community. VUMC reviews module submissions and publishes those that pass review in the REDCap Repo.

As of January 2020, the repo contains 133 published modules. This large and growing collection is the database the EM Framework queries for module location and download. It allows the EM Framework to locate module upgrades suitable for a REDCap version and deliver them quickly and easily to the REDCap server.

> The vetting process performed by the VUMC REDCap team adds value to the modules in the REDCap Repo. The review process covers the basics but also addresses some esoteric and import details. Most important for the immediate need, the module must _work_. It must enable without errors, have a meaningful title and description, and do what it claims to do. The source code for the module must be published for public review in a software repository at [GitHub.com](https://github.com) and it must have an [open source software license](https://en.wikipedia.org/wiki/Open-source_license). The software must embrace coding styles that will make it reasonably easy to support as REDCap evolves. The software should also embrace REDCap's conventions for data access and data update where practicable. The module should follow EM Framework and software industry conventions for describing its software dependencies. The module should follow REDCap and software industry conventions for securing access to data and software features, sanitizing input, and resisting attack. The full details of the current review guidelines can be found at [External Modules: Module Review Guidelines](https://redcap.vanderbilt.edu/consortium/modules/external_modules_review_guidelines.pdf)

Reviewer: ```suggestion
The vetting process performed by the VUMC REDCap team adds value to the modules in the REDCap Repo. The review process covers the basics but also addresses some esoteric and important details. Most importantly for the immediate need, the module must _work_. It must enable without errors, have a meaningful title and description, and do what it claims to do. The source code for the module must be published for public review in a software repository at [GitHub.com](https://github.com) and it must have an [open source software license](https://en.wikipedia.org/wiki/Open-source_license). The software must embrace coding styles that will make it reasonably easy to support as REDCap evolves. The software should also embrace REDCap's conventions for data access and data update where practicable. The module should follow EM Framework and software industry conventions for describing its software dependencies. The module should follow REDCap and software industry conventions for securing access to data and software features, sanitizing input, and resisting attack. The full details of the current review guidelines can be found at [External Modules: Module Review Guidelines](https://redcap.vanderbilt.edu/consortium/modules/external_modules_review_guidelines.pdf)
```
Sentence 2: import -> important  
Sentence 3: important -> importantly

The VUMC team's vetting process should never be considered a guarantee of quality, but it definitely _increases_ and _enables_ quality. The  guidelines have specific details software developers can follow to improve the quality of their product. The guidelines also protect a module user's ability and right to do their own code review, provide feedback, make improvements, and publish those improvements.

## Where to find modules

> The REDCap Repo contains the most REDCap popular. The web interface for the repo is publicly accessible at [https://redcap.vanderbilt.edu/consortium/modules/index.php](https://redcap.vanderbilt.edu/consortium/modules/index.php). The repo's interface allows you to search the database of modules and sort the search results. It provides links to each module's git repo on GitHub.com which can provide details about the state and history of the module.

Reviewer: ```suggestion
The REDCap Repo contains most popular REDCap modules. The web interface for the repo is publicly accessible at [https://redcap.vanderbilt.edu/consortium/modules/index.php](https://redcap.vanderbilt.edu/consortium/modules/index.php). The repo's interface allows you to search the database of modules and sort the search results. It provides links to each module's git repo on GitHub.com which can provide details about the state and history of the module.
```
(Sentence 1) I think this is what you meant to say?

Author: I prefer your wording. :-)

Yet not all modules have been released via the REDCap Repo. Some are not yet ready for publication while others were created to be more site or project specific. A great deal of this development work is happening on Github. Many of those modules can be located by searching Github for specific topics the module development community has been using. Those topics--*redcap-external-module* and _redcap-repo_--can be accessed with these URLs: [https://github.com/topics/redcap-external-module](https://github.com/topics/redcap-external-module), [https://github.com/topics/redcap-repo](https://github.com/topics/redcap-repo).

If you locate a module of interest to you, you can use the [GitHub.com](https://github.com) web interface to explore the module's state and history. A README.md is the first stop to understand a module's capabilities. A CHANGELOG.md should provide a history of the module's development. The commit history can also provide a development history though it is less curated. Browse through open issues to see outstanding problems. Look at closed issues to see corrected problems. New and completed features are often described through issues as well.

> If you want to install a module that is in Github but not the REDCap Repo, access the repo's _releases_ in Github to download a zipped copy of the module. In many cases, the zip file is 100% of that you need to install the module. You can download it, extract it, rename the unzipped folder with a suffix like "\_v1.0.0" using the version number of the module, then copy that whole folder to the _./modules/_ folder of your REDCap server. At that point it is ready to be enable as if it had been downloaded from the REDCap Repo.

Reviewer: ```suggestion
If you want to install a module that is in Github but not the REDCap Repo, access the repo's _releases_ in Github to download a zipped copy of the module. In many cases, the zip file is 100% of that you need to install the module. You can download it, extract it, rename the unzipped folder with a suffix like "\_v1.0.0" using the version number of the module, then copy that whole folder to the `./modules/` folder of your REDCap server. At that point it is ready to be enabled as if it had been downloaded from the REDCap Repo.
```

(Final sentence - 1) use monospace block for modules directory
(Final sentence) enable -> enabled

> If you want to follow the development of a module in GitHub, you can create a GitHub account and then _watch_ the module. A GitHub account will also allow you to create and comment on issues for any module.

Reviewer: ```suggestion
If you want to follow the development of a module in GitHub, you can create a GitHub account and then [_watch_](https://help.github.com/en/github/receiving-notifications-about-activity-on-github/watching-and-unwatching-repositories#watching-a-single-repository) the module. A GitHub account will also allow you to create and comment on issues for any module.
```
Add link to GitHub help page on watching repos

> One potential point of confusion when you view a module from GitHub is that each module visible in Github is in _git repository_, a term often shortened to _git repo_ or just _repo_. This is not the same kind of repo as the REDCap Repo. They just happen to share the same name.

Reviewer: ```suggestion
One potential point of confusion when you view a module from GitHub is that each module visible in Github is in a _git repository_, a term often shortened to _git repo_ or just _repo_. This is not the same kind of repo as the REDCap Repo. They just happen to share the same name.
```

is in _git repository_ -> is in a _git repository_
